File: 1_Arrays/Day_006_Majority_Element_II/Solution.py
class Solution:

    '''
        Approach 1: Using HashMap/Dictionary
        Time Complexity: O(n) for traversing the array and O(nlogn) for sorting the result.
        Space Complexity: O(n) for storing the count of each element in the dictionary.
        
            Function to find the majority elements in the array
            The majority element is the element that appears more than n/3 times in the array.
            The function returns a list of all such elements in sorted order.

            The function uses a dictionary to count the occurrences of each element in the array.
            It then iterates through the dictionary to find elements that appear more than n/3 times.
            Finally, it sorts the result and returns it.
            The function handles the case where the input array is empty by returning an empty list.
            The function also handles the case where there are no majority elements by returning an empty list.
            The function uses a dictionary to store the count of each element in the array.
    '''
    def findMajority(self, arr):
        # Get the length of the array
        n = len(arr)
        # Initialize a dictionary to store the count of each element in the array
        result = {}
        # Initialize an empty list to store the majority elements
        majority = []
        # If the array is empty, return an empty list
        if(n == 0):
            return majority
        # Count the occurrences of each element in the array using a dictionary
        # The dictionary stores the element as the key and its count as the value
        # The dictionary is used to keep track of the count of each element in the array.
        for i in range(n):
            result[arr[i]] = (result.get(arr[i], 0) +1)
        # Iterate through the dictionary to find elements that appear more than n/3 times
        # The dictionary is used to find the majority elements in the array.
        for j in result:
            if(result[j] > (n//3)):
                majority.append(j)
        # Sort the result in ascending order before returning it
        # The result is sorted to ensure that the majority elements are returned in ascending order.
        return sorted(majority)

    ''' 
        Approach 2: Using Boyer-Moore’s Voting Algorithm
        Time Complexity: O(n) for traversing the array and O(1) for sorting the result.
        Space Complexity: O(1) for storing the count of each element in the hashmap.

            Function to find the majority elements in the array
            The majority element is the element that appears more than n/3 times in the array.
            The function returns a list of all such elements in sorted order.
            
            The function uses Boyer-Moore's Voting Algorithm to find the majority elements in the array.
            It maintains two candidates and their counts, and iterates through the array to find the majority elements.
            Finally, it checks the counts of the candidates to determine if they are majority elements.
            The function handles the case where the input array is empty by returning an empty list.
            The function also handles the case where there are no majority elements by returning an empty list.
            The function uses two variables to store the candidates and their counts.
    '''
    def findMajority(self, arr):
        # Get the length of the array
        n = len(arr)
        # Initialize an empty list to store the majority elements
        majority = []
        # If the array is empty, return an empty list
        # The function handles the case where the input array is empty by returning an empty list.
        if(n==0):
            return majority
        # Initialize two variables to store the candidates and their counts
        # The variables are used to keep track of the candidates and their counts.
        e1, e2 = -1, -1
        c1, c2 = 0, 0
        # Iterate through the array to find the majority elements
        # The loop iterates through the array to find the majority elements.
        for i in range(n):
            # The condition checks if the current element is equal to the first candidate.
            if(e1 == arr[i]):
                c1+=1
            # The condition checks if the current element is equal to the second candidate.
            elif(e2 == arr[i]):
                c2+=1
            # The condition checks if the first candidate count is zero and assigns the current element to the first candidate.
            # Increment the count of the first candidate.
            elif(c1==0):
                e1=arr[i]
                c1+=1
            # The condition checks if the second candidate count is zero and assigns the current element to the second candidate.
            # Increment the count of the second candidate.
            elif(c2==0):
                e2=arr[i]
                c2+=1
            # The condition checks if the current element is not equal to either candidate.
            # Decrement the counts of both candidates.
            else:
                c1-=1
                c2-=1
        # After the loop, we have two candidates e1 and e2.
        # We need to check if they are majority elements by counting their occurrences in the array.
        # The counts of the candidates are checked to determine if they are majority elements.
        # Initialize the counts of the candidates to zero
        c1, c2 = 0, 0
        # Count the occurrences of each candidate in the array
        # The loop iterates through the array to count the occurrences of each candidate.
        for i in arr:
            # The condition checks if the current element is equal to the first candidate and increments the count of the first candidate.
            if(e1==i):
                c1+=1
            # The condition checks if the current element is equal to the second candidate and increments the count of the second candidate.
            if(e2==i):
                c2+=1
        # Check if the counts of the candidates are greater than n/3 and add them to the majority list
        if(c1 > (n//3)):
            majority.append(e1)
        if((c2 > (n//3)) and (e1!=e2)):
            majority.append(e2)
        # Sort the result in ascending order before returning it
        # The result is sorted to ensure that the majority elements are returned in ascending order.
        return sorted(majority)

File: 1_Arrays/Day_006_Majority_Element_II/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

class Solution {

    /**
     * @param args
     */
    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {3, 2, 3};
        List<Integer> result = solution.findMajority(nums);
        System.out.println(result); // Output: [3]

        // int[] nums2 = {1, 2, 3, 1, 2, 1};
        int[] nums2 = {2,1,1,9,9,9,9,9,9,9,9,9,5,5,5,5,6,6,6,6,6};
        List<Integer> result2 = solution.findMajorityElements(nums2);
        System.out.println(result2); // Output: [1, 2]
    }


    /**
     * * Approach: HashMap
     * * Function to find the majority elements in the array
     * * The majority element is the element that appears more than n/3 times in the array.
     * * * The function uses a HashMap to count the occurrences of each element in the array.
     * * * It then iterates through the HashMap and adds the elements that appear more than n/3 times to the result list.
     * * * Finally, it returns the result list.
     * 
     * * * * Time Complexity: O(n)
     * * * * Space Complexity: O(n)
     * * * * The function uses a HashMap to store the counts of each element in the array.
     * * * * * The function uses a List to store the majority elements.
     * @param nums
     * @return
     */
    public List<Integer> findMajority(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer> result = new HashMap<>();
        List<Integer> majority = new ArrayList<Integer>();
        
        if(n == 0){
            return majority;
        }

        for(int i=0; i<n;i++){
            result.put(nums[i], result.getOrDefault(nums[i], 0) +1);
        }
        
        for(Map.Entry<Integer,Integer> entry: result.entrySet()){
            if(entry.getValue() > (n/3)){
                majority.add(entry.getKey());
            }
        }
        
        return majority;
    }


    /**
     * * Approach: Boyer-Moore’s Voting Algorithm
     * * * Function to find the majority elements in the array
     * * * The majority element is the element that appears more than n/3 times in the array.
     * * * * The function uses the Boyer-Moore’s Voting Algorithm to find the majority elements in the array.
     * * * * The function uses two variables to keep track of the two candidates for the majority element.
     * * * * * The function iterates through the array and updates the candidates and their counts.
     * * * * * The function then resets the counts and counts the occurrences of the candidates in the array.
     * * * * * * Finally, it adds the candidates to the result list if they are majority elements.
     * * * * * * * The function returns the result list.
     * 
     * * * * * Time Complexity: O(n)
     * * * * * Space Complexity: O(1)
     * * * * * The function uses two variables to keep track of the two candidates for the majority element.
     * * * * * * The function uses a List to store the majority elements.
     * @param nums
     * @return
     */
    public List<Integer> findMajorityElements(int[] nums){
        int n = nums.length;
        List<Integer> majority = new ArrayList<Integer>();
        
        if(n ==0)
            return majority;
        int e1 = -1, e2 =-1;
        int c1 = 0, c2 = 0;
        
        for(int i=0; i<n;i++){
            // Increment Count for Candidate1
            if(e1 == nums[i]){
                c1++;
            }
            // Increment Count for Candidate2
            else if(e2 == nums[i]){
                c2++;
            }
            // New Candidate1 if Count is zero
            else if(c1 == 0){
                e1 = nums[i];
                c1++;
            }
            // New Candidate2 if Count is zero
            else if(c2 == 0){
                e2 = nums[i];
                c2++;
            }
            // Decrement Both counts if neither candidate
            else{
                c1--;
                c2--;
            }
        }
        
        // Reset Counts
        c1=0;
        c2=0;
        // Count the Occurences of candidates
        for(int ele: nums){
            if(e1 == ele) c1++;
            if(e2 == ele) c2++; 
        }
        // Add to result if they are majority elements
        if (c1 > (n / 3)) majority.add(e1);
        if ((c2 > (n / 3)) && (e1 != e2)) majority.add(e2);
        
        if (majority.size() == 2 && majority.get(0) > majority.get(1)) {
            int temp = majority.get(0);
            majority.set(0, majority.get(1));
            majority.set(1, temp);
        }
        return majority;
    }
    
}